### PR TITLE
Add support for buildpack test command

### PR DESCRIPTION
**Pull request by dzaporozhets**
Attempt to add support for `buildpack test` command. 

For https://github.com/gliderlabs/herokuish/issues/257

### CODE REVIEW

**Comment by dzaporozhets** [comment]
@michaelshobbs I am quite new to herokuish and rarely do any bash coding. I am wondering if someone from maintainers can help me turn this into mergeable contribution.

**Comment by dzaporozhets** [comment]
`buildpack-test` is mostly copied code from `buildpack-execute` with small modification.  So far `test-compile` works well but `test` does fail:

```
→ make build-in-docker
→ docker run --rm -v /Users/dzaporozhets/Projects/ruby-sinatra:/tmp/app herokuish:dev /bin/herokuish buildpack test
::: BUILDING APP :::
-----> Ruby app detected
::: PREPARE FOR RUNNING TESTS :::
-----> Setting up Test for Ruby/Rack
-----> Using Ruby version: ruby-2.3.3
-----> Installing dependencies using bundler 1.13.7
       Running: bundle install --without development --path vendor/bundle --binstubs vendor/bundle/bin -j4 --deployment
       Warning: the running version of Bundler (1.13.7) is older than the version that created the lockfile (1.15.3). We suggest you upgrade to the latest version of Bundler by running `gem install bundler`.
       Fetching gem metadata from http://rubygems.org/........
       Fetching version metadata from http://rubygems.org/.
       Installing diff-lcs 1.3
       Installing rake 12.0.0
       Installing rack 1.5.2
       Using bundler 1.13.7
       Installing rspec-support 3.6.0
       Installing rack-protection 1.5.0
       Installing rspec-core 3.6.0
       Installing tilt 1.3.6
       Installing rspec-expectations 3.6.0
       Installing rspec-mocks 3.6.0
       Installing rspec 3.6.0
       Installing sinatra 1.4.2
       Bundle complete! 4 Gemfile dependencies, 12 gems now installed.
       Gems in the group development were not installed.
       Bundled gems are installed into ./vendor/bundle.
       Bundle completed (9.79s)
       Cleaning up the bundler cache.
       Warning: the running version of Bundler (1.13.7) is older than the version that created the lockfile (1.15.3). We suggest you upgrade to the latest version of Bundler by running `gem install bundler`.
-----> Detecting rake tasks

::: RUNNING TESTS :::
-----> Running test: bundle exec rspec
       bash: bundle: command not found
```

I used https://github.com/dzaporozhets/ruby-sinatra as sample ruby app to test functionality

**Comment by michaelshobbs** [comment]
@dzaporozhets thanks for the contribution. we will certainly need some test coverage and documentation updates in order to get this PR merged.

**Comment by dzaporozhets** [comment]
@michaelshobbs sure. At this point, I would like just to make it work. I would appreciate if you could take a look at the code and give me some hints on why it fails on https://github.com/gliderlabs/herokuish/pull/271/files#diff-57c992bbac59903eb3fa272073dc4c8eR222

**Comment by michaelshobbs** [comment]
DISCLAIMER: I am by no means a ruby expert nor do I deploy/run/manage ruby apps.

Based on the output and at the risk of sounding obvious, it's failing due to `bundle: command not found`. If I were you, I'd start by sorting out where (if anywhere) `bundle` exists in the container.

File: include/buildpack.bash [hunk 1]
```diff
@@ -176,3 +176,54 @@ buildpack-execute() {
...
 	shopt -u dotglob nullglob
 }
+
+buildpack-test() {
+	declare desc="Run an application tests using installed buildpacks"
+	ensure-paths
+	[[ "$USER" ]] || randomize-unprivileged
+	buildpack-setup > /dev/null
+
+	if [[ -n "$BUILDPACK_URL" ]]; then
```
Comment (michaelshobbs) [inline include/buildpack.bash]: There is some code duplication here. I'd rather see some of this broken out into a function and get re-used by this and anywhere else we handle buildpack detection.

Comment (dzaporozhets) [inline include/buildpack.bash]: @michaelshobbs yep I will extract this into `select-buildpack` function

File: include/buildpack.bash [hunk 2]
```diff
@@ -176,3 +176,54 @@ buildpack-execute() {
...
+	fi
+	cd "$build_path"
+	unprivileged "$selected_path/bin/test-compile" "$build_path" "$cache_path" "$env_path"
+	unprivileged "$selected_path/bin/test" "$build_path" "$env_path"
+	cd - > /dev/null
+
+	shopt -s dotglob nullglob
+	rm -rf $app_path/*
+	mv $build_path/* $app_path
+	shopt -u dotglob nullglob
```
Comment (dzaporozhets) [inline include/buildpack.bash]: @michaelshobbs I copied 5 lines above from `buildpack-execute`.  Any hints what it used for?

Comment (michaelshobbs) [inline include/buildpack.bash]: This moves the compiled app into the `app_path`. I don't think it's necessary for the testing use case

**Comment by dzaporozhets** [comment]
> based on the output and at the risk of sounding obvious, it's failing due to bundle: command not found. If I were you, I'd start by sorting out where (if anywhere) bundle exists in the container.

@michaelshobbs yeah, sorry my communication was unclear :) Basically my question is next:  Do
I need extra code to execute buildpacks commands one by one or just running `$selected_path/bin/CMD` should be enough.

**Comment by michaelshobbs** [comment]
AFAIK, `$selected_path/bin/CMD` is sufficient. However, since we allow custom buildpacks, we should probably wrap these in a check if the command exists and is executable. If not, pop an error and bail.

**Comment by dzaporozhets** [comment]
I refactored it a bit. Tested on Ruby and Clojure projects. 

So far: 
* `test-compile` works
* `test` does not work

File: include/buildpack.bash (hunk 2 above)
Comment (dzaporozhets) [inline include/buildpack.bash]: cool, thanks

**Comment by dzaporozhets** [comment]
Ok seems `test` does not work because you need to source files from `.profile.d` directory created by `test-compile`.

**Comment by dzaporozhets** [comment]
@michaelshobbs I made it work and did some refactoring to avoid code duplication. Can you please take a look at the code base while I try to figure out how to write tests for it?

**Comment by dzaporozhets** [comment]
Also, one question to decide on: are we ok having `buildpack test` command that runs both `test-compile` and `test` inside? Or we can split those into 3 commands: 

1. `buildpack test-compile`
2. `buildpack test`
3. `buildpack test-with-compile`

For personal use, I don't think separation makes sense but there is a potential for caching between compile and test if you build some heavy used CI based on herokuish.   

@michaelshobbs @markpundsack what do you think?

**Comment by dzaporozhets** [comment]
I think I will proceed with 3 commands suggestion from my comment above. Since buildpack has `test` and `test-compile` commands it might make sense to have same commands in herokuish + 1 command to execute both.

**Comment by markpundsack** [comment]
Thanks, that sounds like a good plan.

**Comment by michaelshobbs** [comment]
From the docs, it seems like the point of the `test` command is to compile the app via the correct buildpack and execute tests. Given that, I would actually fall on the other side and say we should only expose the single `test` command that internally calls the `compile` function and executes the app tests. I don't see the need to expose `test-compile` to the CLI.

File: include/buildpack.bash [hunk 3]
```diff
@@ -151,28 +174,53 @@ buildpack-execute() {
...
-			IFS=$'\n'
-			for var in $config_vars; do
-				echo "export $(echo $var | sed -e 's/=/="/' -e 's/$/"/')" >> "$build_path/.profile.d/00_config_vars.sh"
-			done
-			IFS=$OIFS
-		fi
-	fi
-	cd - > /dev/null
-
+move-build-to-app() {
```
Comment (michaelshobbs) [inline include/buildpack.bash]: let's move this up near `_envfile-parse()` and change the function name to `_move-build-to-app`. the pattern here is unexported functions begin with `_` and are grouped with other unexported functions.

File: include/buildpack.bash [hunk 4]
```diff
@@ -117,6 +117,29 @@ buildpack-setup() {
...
+				echo "export $(echo $var | sed -e 's/=/="/' -e 's/$/"/')" >> "$build_path/.profile.d/00_config_vars.sh"
+			done
+			IFS=$OIFS
+		fi
+	fi
+	cd - > /dev/null
+	move-build-to-app
+}
+
+select-buildpack() {
```
Comment (michaelshobbs) [inline include/buildpack.bash]: similar comment here as with `move-build-to-app`

**Comment by dzaporozhets** [comment]
@michaelshobbs thanks for feedback. I will update MR asap and ping you when done

**Comment by dzaporozhets** [comment]
@michaelshobbs comments addressed. Please take a look

**Comment by dzaporozhets** [comment]
I also updated test apps for Ruby and Clojure and enabled `buildpack test` tests for them.

**Comment by dzaporozhets** [comment]
@michaelshobbs Summary: 
1. Updated code based on your code review
1. Added new method in `buildpacks/test` to test `herokuish buildpack test`
1. Added tests to sample clojure and ruby apps
1. Enable testing of `herokuish buildpack test` for clojure and ruby apps
1. Bumped clojure buildpack version because testing is fixed there.

**Comment by dzaporozhets** [comment]
@michaelshobbs I rebased and added a feature to README and changelog. I see you added @stickler-ci bot for automatic code review. Do I need to address any of those?

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,7 @@ All notable changes to this project will be documented in this file.
 ### Fixed
 
 ### Added
+- @dzaporozhets Add support for `buildpack test` command
```
Comment (michaelshobbs) [inline CHANGELOG.md]: Go ahead and remove this for now. We update the changelog on release.

Comment (dzaporozhets) [inline CHANGELOG.md]: @michaelshobbs got it. fixed

**Comment by michaelshobbs** [comment]
@dzaporozhets rebasing this is sort of a pain. i'm rebasing it locally now and will push up a new branch + PR

**Comment by swistaczek** [comment]
Hey, is it possible to use only 'test-compile' task? I would like to install all dependencies separately and share them in other tasks during CI pipeline.

**Comment by josegonzalez** [comment]
@swistaczek please file a new issue rather than commenting on a year-old pull request. Thank you.

**Pull request closed without merging**